### PR TITLE
Pluggable monitor

### PR DESCRIPTION
**Pull request by AlbyIanna**
### Motivation
The main reason for this PR is to use the latest "pluggable monitor" gRPC APIs #769. But in order to use the new APIs, we needed to refactor all the Serial Monitor architecture. All this work allowed us to solve a number of issues that weren't impossible or too expensive to solve before the change, and we also took advantage of this effort to implement a more solid architecture. 

This PR solves:
- #586
- #769
- #613

### Change description

The architecture has been redesigned like this:
image

#### Call Stack / Event Flow when opening Serial monitor or Plotter

Numbers correspond to those in the diagram above.

1. A monitor is opened (`monitor-widget` or `plotter-frontend-contribution`) from the IDE;
2. `monitor-widget` or `plotter-frontend-contribution` invokes `.startMonitor()` method on `monitor-manager-proxy-client-impl` from `.onBeforeAttach()` or `.startPlotter()`, respectively;
3. `monitor-manager-proxy-client-impl.startMonitor()` invokes `.startMonitor()` method on `monitor-manager-proxy-impl`;
4. `monitor-manager-proxy-impl.startMonitor()` invokes `.startMonitor()` method on `monitor-manager`;
5. `monitor-manager` creates a `monitor-service` from `monitor-service-factory` if not existing, injecting `monitor-settings-provider-impl` and `web-socket-provider-impl`;
6. `monitor-manager` invokes `.start()` method on created/existing `monitor-service`;
    - _this creates a duplex stream (gRPC) to communicate with our CLI;_
    - _An injected WebSocket server is instantiated here;_
7. The result from `monitor-service.start()` method is returned to the `monitor-manager-proxy-impl` when it's duplex connection is successfully connected (if no error);
8. `monitor-manager-proxy-impl` invokes `.connect()` on the `monitor-manager-proxy-client-impl` if value returned from `monitor-service.start()` method is `Status.OK` or `Status.ALREADY_CONNECTED`;
    - _this creates a WebSocket client linked to the WebSocket server created for our `monitor-service`;_

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by fstasi** [comment]
- [ ] reset board does not reconnect
  - [ ] closing the monitor and re-opening fixes the problem
- [ ] missing disconnected status in the serial monitor panel
- [ ] serial plotter connect to the right port only when the serial monitor is open
  - [ ] this is probably due to the fact that the monitor updates the status of the UI with the correct port
- [ ] upload is broken

**Comment by kittaakos** [comment]
I see `selectedBoard.board.fbqn && selectedBoard.port`, `!board || !port`, .etc checks repeatedly and I know why we need this. Still, I propose finding a good name for this "use-case" and single-source the checks and assertion logic somewhere. It will increase the readability of the code.

FYI: for example, there is a concept for an ["available board"](https://github.com/arduino/arduino-ide/blob/522a5c6e0128644a988a8ec671dfaa2f4b30494d/arduino-ide-extension/src/browser/boards/boards-service-provider.ts#L590-L594) (something the IDE2 can reference when verifying code), if the [port is available](https://github.com/arduino/arduino-ide/blob/522a5c6e0128644a988a8ec671dfaa2f4b30494d/arduino-ide-extension/src/browser/boards/boards-service-provider.ts#L616-L620), maybe we can name it as an "attached board", similar.

The generic proposal is to find a good name for the same logic and reuse it if possible.

**Review by ubidefeo — APPROVED** [review]
Tested and working.
LGTM

**Comment by ubidefeo** [comment]
@PaulStoffregen 
would you mind giving the build from this PR a run?
I'm having pretty good results with your experimental Teensy Platform :)

**Comment by ubidefeo** [comment]
@PaulStoffregen 
builds are at the bottom of this page, let me know if you can't see them
https://github.com/arduino/arduino-ide/actions/runs/2447002574

**Comment by PaulStoffregen** [comment]
@ubidefeo - I'm running it now on my Linux desktop machine.  Seems to be working pretty well.  I'm happy to see it's handling Teensy 4.0 maximum speed printing pretty well, with 2 processes each using about 70% of a CPU core, and 4 others adding up to about 55%.

image

**Comment by PaulStoffregen** [comment]
Running a Teensy 4.0 overclocked to 912 MHz, and the pluggable monitor is managing 785000 lines/sec.  Very respectable speed.

I have some rough edges to fix in my package's uploading utility.  Looks like the IDE has some minor serial monitor bugs, probably unrelated to pluggable monitor.

File: arduino-ide-extension/src/browser/arduino-ide-frontend-module.ts [hunk 1]
```diff
@@ -523,7 +540,9 @@ export default new ContainerModule((bind, unbind, isBound, rebind) => {
 
   bind(SearchInWorkspaceWidget).toSelf();
   rebind(TheiaSearchInWorkspaceWidget).toService(SearchInWorkspaceWidget);
-  
+
+  rebind(TheiaEditorManager).to(EditorManager);
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/arduino-ide-frontend-module.ts:544]: Please check this. Most likely, it's an incorrect manual merge.

File: arduino-ide-extension/package.json [hunk 2]
```diff
@@ -155,7 +155,7 @@
   ],
   "arduino": {
     "cli": {
-      "version": "0.23.0"
+      "version": "0.23.0-rc1"
```
Comment (kittaakos) [inline arduino-ide-extension/package.json:158]: Should be without `-rc1`.

**Pull request merged**